### PR TITLE
Fixed #24693 -- Added label and label_lower property to Options

### PR DESCRIPTION
**Pull request by delgiudices**
Ticket : https://code.djangoproject.com/ticket/24693

### CODE REVIEW

**Comment by MarkusH** [comment]
Thanks for the pull request :)

Can you go through the Django source code and replace the places where either `(app_label, model_name)` or `(app_label, object_name)` is used with the respective label.

**Comment by delgiudices** [comment]
Will do

File: django/core/serializers/base.py [hunk 1]
```diff
@@ -163,8 +163,8 @@ def __init__(self, obj, m2m_data=None):
         self.m2m_data = m2m_data
 
     def __repr__(self):
-        return "<DeserializedObject: %s.%s(pk=%s)>" % (
-            self.object._meta.app_label, self.object._meta.object_name, self.object.pk)
+        return "<DeserializedObject: %s(pk=%s)>" % (
```
Comment (MarkusH) [inline django/core/serializers/base.py]: this changes the repr to `<DeserializedObject: (foo.Bar)(pk=%s)>` I suspect

**Comment by timgraham** [comment]
buildbot, add to whitelist.

File: django/db/models/options.py [hunk 2]
```diff
@@ -177,6 +177,14 @@ def _map_model_details(self, link):
         return link, model, direct, m2m
```
Comment (timgraham) [inline django/db/models/options.py]: There's `model_label = '%s.%s' % (self.app_label, self.model_name)` in this file which could use `self.label_lower`

File: docs/ref/models/options.txt [hunk 3]
```diff
@@ -29,6 +29,10 @@ Available ``Meta`` options
 
         app_label = 'myapp'
 
+    .. versionadded:: 1.9
```
Comment (timgraham) [inline docs/ref/models/options.txt]: Let's add a section to the bottom of this file called something like "Read-only ``Meta`` attributes".

Then we can follow the same format of the other options with a heading, `.. attribute:: Options.label`, and a description.

File: docs/ref/models/options.txt [hunk 4]
```diff
@@ -29,6 +29,10 @@ Available ``Meta`` options
...
+
+    If you want to represent a model with the format ``app_label.object_name`` or ``app_label.model_name`` you can use ``model._meta.label`` or ``model._meta.label_lower`` respectively.
```
Comment (timgraham) [inline docs/ref/models/options.txt]: Please wrap docs at 79 characters.

File: tests/model_meta/results.py [hunk 5]
```diff
@@ -791,4 +791,17 @@
...
+        BasePerson: 'model_meta.BasePerson',
+        Person: 'model_meta.Person',
+        Relating: 'model_meta.Relating',
+    },
+    'lower_labels': {
+        AbstractPerson: 'model_meta.abstractperson',
+        BasePerson: 'model_meta.baseperson',
+        Person: 'model_meta.person',
+        Relating: 'model_meta.relating',
+    }
```
Comment (timgraham) [inline tests/model_meta/results.py]: including trailing comma and chop blank line that follows

**Comment by timgraham** [comment]
After you update the patch, please uncheck "Patch needs improvement" on the ticket so it appears in the review queue, thanks!

**Comment by delgiudices** [comment]
I didn't know about "Patch needs improvement" will keep in mind for future PRs thanks

File: django/db/models/manager.py [hunk 6]
```diff
@@ -77,8 +77,7 @@ def __init__(self):
     def __str__(self):
         """ Return "app_label.model_label.manager_name". """
         model = self.model
-        app = model._meta.app_label
-        return '%s.%s.%s' % (app, model._meta.object_name, self.name)
+        return '%s.%s' % (model._meta.label, self.name)
```
Comment (MarkusH) [inline django/db/models/manager.py]: You can remove the `model = self.model` above and use `self.model._meta.label` here.

File: docs/ref/models/options.txt [hunk 7]
```diff
@@ -397,3 +403,20 @@ Django quotes column and table names behind the scenes.
...
+
+Read-only ``Meta`` attributes
+=============================
+
+``label``
+---------
+
+.. attribute:: Options.label
+
+    Representation of the object, returns ``app_label.object_name``.
```
Comment (timgraham) [inline docs/ref/models/options.txt]: need .. versionadded:: 1.9 annotations for these attributes

Comment (timgraham) [inline docs/ref/models/options.txt]: maybe we could make `app_label.object_name` links to the respective attributes and also give an example like 'polls.Question'

File: docs/ref/models/options.txt (hunk 3 above)
Comment (timgraham) [inline docs/ref/models/options.txt]: I think we can remove this paragraph here.

File: docs/ref/models/options.txt (hunk 7 above)
Comment (timgraham) [inline docs/ref/models/options.txt]: Actually we need docs for `object_name` and `model_name` in this section as well.

Comment (MarkusH) [inline docs/ref/models/options.txt]: This should do the job, @delgiudices 

```rst
.. attribute:: Options.label

    .. versionadded:: 1.9

    Representation of the object, returns ``app_label.object_name``.
```

**Comment by MarkusH** [comment]
```diff
diff --git a/django/db/migrations/state.py b/django/db/migrations/state.py
index 7c4984f..4b24414 100644
--- a/django/db/migrations/state.py
+++ b/django/db/migrations/state.py
@@ -424,7 +424,7 @@ class ModelState(object):
         # Make our record
         bases = tuple(
             (
-                "%s" % base._meta.label_lower
+                base._meta.label_lower
                 if hasattr(base, "_meta") else
                 base
             )
diff --git a/django/db/models/base.py b/django/db/models/base.py
index 42474c5..27764bb 100644
--- a/django/db/models/base.py
+++ b/django/db/models/base.py
@@ -327,7 +327,6 @@ class ModelBase(type):
             cls.__doc__ = "%s(%s)" % (cls.__name__, ", ".join(f.name for f in opts.fields))

         get_absolute_url_override = settings.ABSOLUTE_URL_OVERRIDES.get(opts.label_lower)
-
         if get_absolute_url_override:
             setattr(cls, 'get_absolute_url', get_absolute_url_override)

diff --git a/django/forms/models.py b/django/forms/models.py
index 5828b0a..1ea0dea 100644
--- a/django/forms/models.py
+++ b/django/forms/models.py
@@ -973,12 +973,12 @@ def _get_foreign_key(parent_model, model, fk_name=None, can_fail=False):
                     (fk.remote_field.model != parent_model and
                      fk.remote_field.model not in parent_model._meta.get_parent_list()):
                 raise ValueError(
-                    "fk_name '%s' is not a ForeignKey to '%s'."
-                    % (fk_name, parent_model._meta.label))
+                    "fk_name '%s' is not a ForeignKey to '%s'." % (fk_name, parent_model._meta.label)
+                )
         elif len(fks_to_parent) == 0:
             raise ValueError(
-                "'%s' has no field named '%s'."
-                % (model._meta.label, fk_name))
+                "'%s' has no field named '%s'." % (model._meta.label, fk_name)
+            )
     else:
         # Try to discover what the ForeignKey from model to parent_model is
         fks_to_parent = [
diff --git a/docs/ref/models/options.txt b/docs/ref/models/options.txt
index 2098395..1dff215 100644
--- a/docs/ref/models/options.txt
+++ b/docs/ref/models/options.txt
@@ -414,7 +414,8 @@ Read-only ``Meta`` attributes

     .. versionadded:: 1.9

-    Representation of the object, returns ``app_label.object_name``.
+    Representation of the object, returns ``app_label.object_name``, e.g.
+    ``'polls.Question'``.

 ``label_lower``
 ---------------
@@ -423,4 +424,5 @@ Read-only ``Meta`` attributes

     .. versionadded:: 1.9

-    Representation of the model, returns ``app_label.model_name``.
+    Representation of the model, returns ``app_label.model_name``, e.g.
+    ``'polls.question'``.
```

**Comment by MarkusH** [comment]
Merged in 69ddc1b3da043195a0f4e09211d395724b42c70b. Thanks! :smile:

**Pull request closed without merging**